# Information Retrieval 1#
## Assignment 2: Retrieval models [100 points] ##

In this assignment you will get familiar with basic and advanced information retrieval concepts. You will implement different information retrieval ranking models and evaluate their performance.

We provide you with a Indri index. To query the index, you'll use a Python package ([pyndri](https://github.com/cvangysel/pyndri)) that allows easy access to the underlying document statistics.

For evaluation you'll use the [TREC Eval](https://github.com/usnistgov/trec_eval) utility, provided by the National Institute of Standards and Technology of the United States. TREC Eval is the de facto standard way to compute Information Retrieval measures and is frequently referenced in scientific papers.

This is a **groups-of-three assignment**, the deadline is **Wednesday, January 31st**. Code quality, informative comments and convincing analysis of the results will be considered when grading. Submission should be done through blackboard, questions can be asked on the course [Piazza](piazza.com/university_of_amsterdam/spring2018/52041inr6y/home).

### Technicalities (must-read!) ###

The assignment directory is organized as follows:
   * `./assignment.ipynb` (this file): the description of the assignment.
   * `./index/`: the index we prepared for you.
   * `./ap_88_90/`: directory with ground-truth and evaluation sets:
      * `qrel_test`: test query relevance collection (**test set**).
      * `qrel_validation`: validation query relevance collection (**validation set**).
      * `topics_title`: semicolon-separated file with query identifiers and terms.

You will need the following software packages (tested with Python 3.5 inside [Anaconda](https://conda.io/docs/user-guide/install/index.html)):
   * Python 3.5 and Jupyter
   * Indri + Pyndri (Follow the installation instructions [here](https://github.com/nickvosk/pyndri/blob/master/README.md))
   * gensim [link](https://radimrehurek.com/gensim/install.html)
   * TREC Eval [link](https://github.com/usnistgov/trec_eval)

### TREC Eval primer ###
The TREC Eval utility can be downloaded and compiled as follows:

    git clone https://github.com/usnistgov/trec_eval.git
    cd trec_eval
    make

TREC Eval computes evaluation scores given two files: ground-truth information regarding relevant documents, named *query relevance* or *qrel*, and a ranking of documents for a set of queries, referred to as a *run*. The *qrel* will be supplied by us and should not be changed. For every retrieval model (or combinations thereof) you will generate a run of the top-1000 documents for every query. The format of the *run* file is as follows:

    $query_identifier Q0 $document_identifier $rank_of_document_for_query $query_document_similarity $run_identifier
    
where
   * `$query_identifier` is the unique identifier corresponding to a query (usually this follows a sequential numbering).
   * `Q0` is a legacy field that you can ignore.
   * `$document_identifier` corresponds to the unique identifier of a document (e.g., APXXXXXXX where AP denotes the collection and the Xs correspond to a unique numerical identifier).
   * `$rank_of_document_for_query` denotes the rank of the document for the particular query. This field is ignored by TREC Eval and is only maintained for legacy support. The ranks are computed by TREC Eval itself using the `$query_document_similarity` field (see next). However, it remains good practice to correctly compute this field.
   * `$query_document_similarity` is a score indicating the similarity between query and document where a higher score denotes greater similarity.
   * `$run_identifier` is an identifier of the run. This field is for your own convenience and has no purpose beyond bookkeeping.
   
For example, say we have two queries: `Q1` and `Q2` and we rank three documents (`DOC1`, `DOC2`, `DOC3`). For query `Q1`, we find the following similarity scores `score(Q1, DOC1) = 1.0`, `score(Q1, DOC2) = 0.5`, `score(Q1, DOC3) = 0.75`; and for `Q2`: `score(Q2, DOC1) = -0.1`, `score(Q2, DOC2) = 1.25`, `score(Q1, DOC3) = 0.0`. We can generate run using the following snippet:

In [338]:
import logging
import sys
import os

def write_run(model_name, data, out_f,
              max_objects_per_query=sys.maxsize,
              skip_sorting=False):
    """
    Write a run to an output file.
    Parameters:
        - model_name: identifier of run.
        - data: dictionary mapping topic_id to object_assesments;
            object_assesments is an iterable (list or tuple) of
            (relevance, object_id) pairs.
            The object_assesments iterable is sorted by decreasing order.
        - out_f: output file stream.
        - max_objects_per_query: cut-off for number of objects per query.
    """

    for subject_id, object_assesments in data.items():
        if not object_assesments:
            logging.warning('Received empty ranking for %s; ignoring.',
                            subject_id)

            continue

        # Probe types, to make sure everything goes alright.
        # assert isinstance(object_assesments[0][0], float) or \
        #     isinstance(object_assesments[0][0], np.float32)
        assert isinstance(object_assesments[0][1], str) or \
            isinstance(object_assesments[0][1], bytes)

        if not skip_sorting:
            object_assesments = sorted(object_assesments, reverse=True)

        if max_objects_per_query < sys.maxsize:
            object_assesments = object_assesments[:max_objects_per_query]

        if isinstance(subject_id, bytes):
            subject_id = subject_id.decode('utf8')

        for rank, (relevance, object_id) in enumerate(object_assesments):
            if isinstance(object_id, bytes):
                object_id = object_id.decode('utf8')

            out_f.write(
                '{subject} Q0 {object} {rank} {relevance} '
                '{model_name}\n'.format(
                    subject=subject_id,
                    object=object_id,
                    rank=rank + 1,
                    relevance=relevance,
                    model_name=model_name))
            
# The following writes the run to standard output.
# In your code, you should write the runs to local
# storage in order to pass them to trec_eval.
write_run(
    model_name='example',
    data={
        'Q1': ((1.0, 'DOC1'), (0.5, 'DOC2'), (0.75, 'DOC3')),
        'Q2': ((-0.1, 'DOC1'), (1.25, 'DOC2'), (0.0, 'DOC3')),
    },
    out_f=sys.stdout,
    max_objects_per_query=1000)

Q1 Q0 DOC1 1 1.0 example
Q1 Q0 DOC3 2 0.75 example
Q1 Q0 DOC2 3 0.5 example
Q2 Q0 DOC2 1 1.25 example
Q2 Q0 DOC3 2 0.0 example
Q2 Q0 DOC1 3 -0.1 example


Now, imagine that we know that `DOC1` is relevant and `DOC3` is non-relevant for `Q1`. In addition, for `Q2` we only know of the relevance of `DOC3`. The query relevance file looks like:

    Q1 0 DOC1 1
    Q1 0 DOC3 0
    Q2 0 DOC3 1
    
We store the run and qrel in files `example.run` and `example.qrel` respectively on disk. We can now use TREC Eval to compute evaluation measures. In this example, we're only interested in Mean Average Precision and we'll only show this below for brevity. However, TREC Eval outputs much more information such as NDCG, recall, precision, etc.

    $ trec_eval -m all_trec -q example.qrel example.run | grep -E "^map\s"
    > map                   	Q1	1.0000
    > map                   	Q2	0.5000
    > map                   	all	0.7500
    
Now that we've discussed the output format of rankings and how you can compute evaluation measures from these rankings, we'll now proceed with an overview of the indexing framework you'll use.

### Pyndri primer ###
For this assignment you will use [Pyndri](https://github.com/cvangysel/pyndri) [[1](https://arxiv.org/abs/1701.00749)], a python interface for [Indri](https://www.lemurproject.org/indri.php). We have indexed the document collection and you can query the index using Pyndri. We will start by giving you some examples of what Pyndri can do:

First we read the document collection index with Pyndri:

In [339]:
import pyndri

index = pyndri.Index('index/')

The loaded index can be used to access a collection of documents in an easy manner. We'll give you some examples to get some idea of what it can do, it is up to you to figure out how to use it for the remainder of the assignment.

First let's look at the number of documents, since Pyndri indexes the documents using incremental identifiers we can simply take the lowest index and the maximum document and consider the difference:

In [340]:
print(index.maximum_document())
print(index.document_base())

print("There are %d documents in this collection." % (index.maximum_document() - index.document_base()))

164598
1
There are 164597 documents in this collection.


Let's take the first document out of the collection and take a look at it:

In [341]:
# APXXXXXXXXX corresponds to the unique identifier of a document:
# e.g., APXXXXXXX where AP denotes the collection and the Xs correspond to a unique numerical identifier.

example_document = index.document(1)
print(example_document)
print(len(example_document[1]))

('AP890425-0001', (1360, 192, 363, 0, 880, 0, 200, 0, 894, 412, 92160, 3, 192, 0, 363, 34, 1441, 0, 174134, 0, 200, 0, 894, 412, 2652, 0, 810, 107, 49, 4903, 420, 0, 1, 48, 35, 489, 0, 35, 687, 192, 243, 0, 249311, 1877, 0, 1651, 1174, 0, 2701, 117, 412, 0, 810, 391, 245233, 1225, 5838, 16, 0, 233156, 3496, 0, 393, 17, 0, 2435, 4819, 930, 0, 0, 200, 0, 894, 0, 22, 398, 145, 0, 3, 271, 115, 0, 1176, 2777, 292, 0, 725, 192, 0, 0, 50046, 0, 1901, 1130, 0, 192, 0, 408, 0, 243779, 0, 0, 553, 192, 0, 363, 0, 3747, 0, 0, 0, 0, 1176, 0, 1239, 0, 0, 1115, 17, 0, 0, 585, 192, 1963, 0, 0, 412, 54356, 0, 773, 0, 0, 0, 192, 0, 0, 1130, 0, 363, 0, 545, 192, 0, 1174, 1901, 1130, 0, 4, 398, 145, 39, 0, 577, 0, 355, 0, 491, 0, 6025, 0, 0, 193156, 88, 34, 437, 0, 0, 1852, 0, 828, 0, 1588, 0, 0, 0, 2615, 0, 0, 107, 49, 420, 0, 0, 190, 7, 714, 2701, 0, 237, 192, 157, 0, 412, 34, 437, 0, 0, 200, 6025, 26, 0, 0, 0, 0, 363, 0, 22, 398, 145, 0, 200, 638, 126222, 6018, 0, 880, 0, 0, 161, 0, 0, 319, 894, 2701, 

Here we see a document consists of two things, a string representing the external document identifier and an integer list representing the identifiers of words that make up the document. Pyndri uses integer representations for words or terms, thus a token_id is an integer that represents a word whereas the token is the actual text of the word/term. Every id has a unique token and vice versa with the exception of stop words: words so common that there are uninformative, all of these receive the zero id.

To see what some ids and their matching tokens we take a look at the dictionary of the index:

In [342]:
token2id, id2token, _ = index.get_dictionary()
print(list(id2token.items())[:15])
print(len(id2token))

[(1, 'new'), (2, 'percent'), (3, 'two'), (4, '1'), (5, 'people'), (6, 'million'), (7, '000'), (8, 'government'), (9, 'president'), (10, 'years'), (11, 'state'), (12, '2'), (13, 'states'), (14, 'three'), (15, 'time')]
267318


Using this dictionary we can see the tokens for the (non-stop) words in our example document:

In [343]:
print([id2token[word_id] for word_id in example_document[1] if word_id > 0])

['52', 'students', 'arrested', 'takeover', 'university', 'massachusetts', 'building', 'fifty', 'two', 'students', 'arrested', 'tuesday', 'evening', 'occupying', 'university', 'massachusetts', 'building', 'overnight', 'protest', 'defense', 'department', 'funded', 'research', 'new', 'york', 'city', 'thousands', 'city', 'college', 'students', 'got', 'unscheduled', 'holiday', 'demonstrators', 'occupied', 'campus', 'administration', 'building', 'protest', 'possible', 'tuition', 'increases', 'prompting', 'officials', 'suspend', 'classes', '60', 'police', 'riot', 'gear', 'arrived', 'university', 'massachusetts', '5', 'p', 'm', 'two', 'hours', 'later', 'bus', 'drove', 'away', '29', 'students', 'camped', 'memorial', 'hall', 'students', 'charged', 'trespassing', '23', 'students', 'arrested', 'lying', 'bus', 'prevent', 'leaving', 'police', '300', 'students', 'stood', 'building', 'chanting', 'looking', 'students', 'hall', 'arrested', '35', 'students', 'occupied', 'memorial', 'hall', '1', 'p', 'm',

The reverse can also be done, say we want to look for news about the "University of Massachusetts", the tokens of that query can be converted to ids using the reverse dictionary:

In [344]:
query_tokens = index.tokenize("University of Massachusetts")
print("Query by tokens:", query_tokens)
query_id_tokens = [token2id.get(query_token,0) for query_token in query_tokens]
print("Query by ids with stopwords:", query_id_tokens)
query_id_tokens = [word_id for word_id in query_id_tokens if word_id > 0]
print("Query by ids without stopwords:", query_id_tokens)

Query by tokens: ['university', '', 'massachusetts']
Query by ids with stopwords: [200, 0, 894]
Query by ids without stopwords: [200, 894]


Naturally we can now match the document and query in the id space, let's see how often a word from the query occurs in our example document:

In [345]:
matching_words = sum([True for word_id in example_document[1] if word_id in query_id_tokens])
print("Document %s has %d word matches with query: \"%s\"." % (example_document[0], matching_words, ' '.join(query_tokens)))
print("Document %s and query \"%s\" have a %.01f%% overlap." % (example_document[0], ' '.join(query_tokens),matching_words/float(len(example_document[1]))*100))

Document AP890425-0001 has 13 word matches with query: "university  massachusetts".
Document AP890425-0001 and query "university  massachusetts" have a 2.5% overlap.


While this is certainly not everything Pyndri can do, it should give you an idea of how to use it. Please take a look at the [examples](https://github.com/cvangysel/pyndri) as it will help you a lot with this assignment.

**CAUTION**: Avoid printing out the whole index in this Notebook as it will generate a lot of output and is likely to corrupt the Notebook.

### Parsing the query file
You can parse the query file (`ap_88_89/topics_title`) using the following snippet:

In [346]:
import collections
import io
import logging
import sys

def parse_topics(file_or_files,
                 max_topics=sys.maxsize, delimiter=';'):
    assert max_topics >= 0 or max_topics is None

    topics = collections.OrderedDict()

    if not isinstance(file_or_files, list) and \
            not isinstance(file_or_files, tuple):
        if hasattr(file_or_files, '__iter__'):
            file_or_files = list(file_or_files)
        else:
            file_or_files = [file_or_files]

    for f in file_or_files:
        assert isinstance(f, io.IOBase)

        for line in f:
            assert(isinstance(line, str))

            line = line.strip()

            if not line:
                continue

            topic_id, terms = line.split(delimiter, 1)

            if topic_id in topics and (topics[topic_id] != terms):
                    logging.error('Duplicate topic "%s" (%s vs. %s).',
                                  topic_id,
                                  topics[topic_id],
                                  terms)

            topics[topic_id] = terms

            if max_topics > 0 and len(topics) >= max_topics:
                print("Max capacity reached... Breaking...")
                break

    return topics

with open('./ap_88_89/topics_title', 'r') as f_topics:
    print(parse_topics([f_topics]))

OrderedDict([('51', 'Airbus Subsidies'), ('52', 'South African Sanctions'), ('53', 'Leveraged Buyouts'), ('54', 'Satellite Launch Contracts'), ('55', 'Insider Trading'), ('56', 'Prime (Lending) Rate Moves, Predictions'), ('57', 'MCI'), ('58', 'Rail Strikes'), ('59', 'Weather Related Fatalities'), ('60', 'Merit-Pay vs. Seniority'), ('61', 'Israeli Role in Iran-Contra Affair'), ('62', "Military Coups D'etat"), ('63', 'Machine Translation'), ('64', 'Hostage-Taking'), ('65', 'Information Retrieval Systems'), ('66', 'Natural Language Processing'), ('67', 'Politically Motivated Civil Disturbances'), ('68', 'Health Hazards from Fine-Diameter Fibers'), ('69', 'Attempts to Revive the SALT II Treaty'), ('70', 'Surrogate Motherhood'), ('71', 'Border Incursions'), ('72', 'Demographic Shifts in the U.S.'), ('73', 'Demographic Shifts across National Boundaries'), ('74', 'Conflicting Policy'), ('75', 'Automation'), ('76', 'U.S. Constitution - Original Intent'), ('77', 'Poaching'), ('78', 'Greenpeace'

### Task 1: Implement and compare lexical IR methods [35 points] ### 

In this task you will implement a number of lexical methods for IR using the **Pyndri** framework. Then you will evaluate these methods on the dataset we have provided using **TREC Eval**.

Use the **Pyndri** framework to get statistics of the documents (term frequency, document frequency, collection frequency; **you are not allowed to use the query functionality of Pyndri**) and implement the following scoring methods in **Python**:

- [TF-IDF](http://nlp.stanford.edu/IR-book/html/htmledition/tf-idf-weighting-1.html) and 
- [BM25](http://nlp.stanford.edu/IR-book/html/htmledition/okapi-bm25-a-non-binary-model-1.html) with k1=1.2 and b=0.75. **[5 points]**
- Language models ([survey](https://drive.google.com/file/d/0B-zklbckv9CHc0c3b245UW90NE0/view))
    - Jelinek-Mercer (explore different values of 𝛌 in the range [0.1, 0.5, 0.9]). **[5 points]**
    - Dirichlet Prior (explore different values of 𝛍 [500, 1000, 1500]). **[5 points]**
    - Absolute discounting (explore different values of 𝛅 in the range [0.1, 0.5, 0.9]). **[5 points]**
    - [Positional Language Models](http://sifaka.cs.uiuc.edu/~ylv2/pub/sigir09-plm.pdf) define a language model for each position of a document, and score a document based on the scores of its PLMs. The PLM is estimated based on propagated counts of words within a document through a proximity-based density function, which both captures proximity heuristics and achieves an effect of “soft” passage retrieval. Implement the PLM, all five kernels, but only the Best position strategy to score documents. Use 𝛔 equal to 50, and Dirichlet smoothing with 𝛍 optimized on the validation set (decide how to optimize this value yourself and motivate your decision in the report). **[10 points]**
    
Implement the above methods and report evaluation measures (on the test set) using the hyper parameter values you optimized on the validation set (also report the values of the hyper parameters). Use TREC Eval to obtain the results and report on `NDCG@10`, Mean Average Precision (`MAP@1000`), `Precision@5` and `Recall@1000`.

For the language models, create plots showing `NDCG@10` with varying values of the parameters. You can do this by chaining small scripts using shell scripting (preferred) or execute trec_eval using Python's `subprocess`.

Compute significance of the results using a [two-tailed paired Student t-test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html) **[5 points]**. Be wary of false rejection of the null hypothesis caused by the [multiple comparisons problem](https://en.wikipedia.org/wiki/Multiple_comparisons_problem). There are multiple ways to mitigate this problem and it is up to you to choose one.

Analyse the results by identifying specific queries where different methods succeed or fail and discuss possible reasons that cause these differences. This is *very important* in order to understand how the different retrieval functions behave.

**NOTE**: Don’t forget to use log computations in your calculations to avoid underflows. 

**IMPORTANT**: You should structure your code around the helper functions we provide below.

In [347]:
import time

start_time = time.time()

with open('./ap_88_89/topics_title', 'r') as f_topics:
    queries = parse_topics([f_topics])

index = pyndri.Index('index/')

num_documents = index.maximum_document() - index.document_base()

print('Total number of documents: ', num_documents)

dictionary = pyndri.extract_dictionary(index)

tokenized_queries = {
    query_id: [dictionary.translate_token(token)
               for token in index.tokenize(query_string)
               if dictionary.has_token(token)]
    for query_id, query_string in queries.items()}

query_term_ids = set(
    query_term_id
    for query_term_ids in tokenized_queries.values()
    for query_term_id in query_term_ids)

#print('Query term IDs: ', query_term_ids)

print('Gathering statistics about', len(query_term_ids), 'terms.')

# inverted index creation.

document_lengths = {}
unique_terms_per_document = {}

inverted_index = collections.defaultdict(dict)
collection_frequencies = collections.defaultdict(int)

total_terms = 0

for int_doc_id in range(index.document_base(), index.maximum_document()):
    ext_doc_id, doc_token_ids = index.document(int_doc_id)

    document_bow = collections.Counter(
        token_id for token_id in doc_token_ids
        if token_id > 0)
    document_length = sum(document_bow.values())

    document_lengths[int_doc_id] = document_length
    total_terms += document_length

    unique_terms_per_document[int_doc_id] = len(document_bow)

    for query_term_id in query_term_ids:
        assert query_term_id is not None

        document_term_frequency = document_bow.get(query_term_id, 0)

        if document_term_frequency == 0:
            continue

        collection_frequencies[query_term_id] += document_term_frequency
        inverted_index[query_term_id][int_doc_id] = document_term_frequency

avg_doc_length = total_terms / num_documents

print('Inverted index creation took', time.time() - start_time, 'seconds.')

Total number of documents:  164597
Gathering statistics about 456 terms.
Inverted index creation took 72.38229584693909 seconds.


In [348]:
print('Total number of queries: ', len(queries))
print('Average document length: ', avg_doc_length)

Total number of queries:  150
Average document length:  256.4381975370147


In [349]:
query_tokens = index.tokenize("Beachfront")
print("Query by tokens:", query_tokens)
query_id_tokens = [token2id.get(query_token,0) for query_token in query_tokens]
print("Query by ids with stopwords:", query_id_tokens)
query_id_tokens = [word_id for word_id in query_id_tokens if word_id > 0]
print("Query by ids without stopwords:", query_id_tokens)
print(inverted_index[query_id_tokens[0]][649])

Query by tokens: ['beachfront']
Query by ids with stopwords: [36062]
Query by ids without stopwords: [36062]
1


In [351]:
import numpy as np

"""Helper functions"""

def term_frequency(term_id, document_id):
    """
    Retrieves the term frequency a given word in a given document
    
    NB: depends on global variable 'inverted_index'
    
    :param term_id: The id of the word in question
    :param document_id: The internal document_id of the document in question
    """
    
    return inverted_index[term_id].get(document_id, 0)

def compute_BM25(doc_length, tf_raw, term_id, k, b):
    """
    Helper function to isolate computations shared by both query and document versions of BM25 method
    """
    
    inv_norm = doc_length / avg_doc_length
    document_freq = len(inverted_index[term_id])
    
    tf_normalized = tf_raw * (avg_doc_length / doc_length)

    # Note: slides are not particularly clear about whether to still use the normalized TF or not.
    # In this case we will continue with this normalized variable.
    tf = ((1.0 + k) * tf_normalized) / (k * ((b - 1.0) + b * (inv_norm)) + tf_normalized)
    
    idf = np.log(num_documents / document_freq)
    
    return tf * idf

def build_document_vector(document_id, vocabulary_size):
    '''
    Creates a vector of length V with the term frequency for each word in the vocabulary at its
    corresponding index
    
    :param document_id: The internal document id
    :param vocabulary_size: The number of words in the vocabulary
    '''
    
    vector = np.zeros(vocabulary_size)
    
    for token_id in inverted_index:
        documents_for_term = inverted_index[token_id]
        
        if document_id in documents_for_term:
            vector[token_id] = documents_for_term[document_id]
        
    return vector
        
def build_query_vector(query_terms, vocabulary_size):
    '''
    Creates a vector of length V with the term frequency for each word in the vocabulary at its
    corresponding index
    
    :param query_terms: A list of query term ids
    :param vocabulary_size: The number of words in the vocabulary
    '''
    
    vector = np.zeros(vocabulary_size)
    counts = collections.Counter(query_terms)
    
    for token_id in counts:
        vector[token_id] = counts[token_id]
        
    return vector

# Vector representation of the collection frequencies dictionary
collection_freq_matrix = np.zeros(len(id2token) + 1)
for key in sorted(collection_frequencies.keys()):
    if key > 0:
        collection_freq_matrix[key] = collection_frequencies[key]

print(collection_freq_matrix[:100])
print(collection_freq_matrix.shape)
# print(collection_freq_matrix.getnnz() - len(collection_freq_matrix.nonzero()[1]))

[      0.  171350.       0.       0.       0.       0.       0.       0.
  119771.       0.       0.   96612.       0.   84131.       0.       0.
       0.       0.       0.   73013.   71782.   69256.       0.       0.
       0.       0.       0.       0.       0.       0.       0.       0.
       0.   56945.       0.   55709.       0.       0.   54180.   53986.
       0.       0.       0.       0.       0.       0.       0.       0.
       0.       0.       0.       0.       0.       0.   46470.   46166.
       0.   45936.       0.       0.       0.       0.       0.       0.
       0.       0.       0.       0.   41803.   41350.       0.       0.
   40954.       0.       0.   40696.       0.   40579.   40331.       0.
   40019.       0.   39433.       0.       0.   38950.       0.       0.
       0.       0.       0.       0.   37135.       0.   36935.       0.
       0.   36376.       0.       0.]
(267319,)


### The Scoring Functions

In [397]:
from sklearn.metrics.pairwise import cosine_similarity
from scipy import sparse
from scipy.stats import norm
import sys

def score_similarity(query_terms, document_id, weight_fn = None, query_weight_fn = None):
    '''
    Scoring function for a document and a query
    
    :param query_terms: A list of query term ids
    :param document_id: The internal document id
    :param weight: A function to re-weight the vectors
    
    returns a score (float)
    '''
    
    vocab_size = len(id2token)
    
    # Create query vector
    # Create document vector
    doc_vec = build_document_vector(document_id, vocab_size)
    query_vec = build_query_vector(query_terms, vocab_size)
    
    # Combine query and document into a matrix with sparse representation (only save indices of non-zero entries)
    paired_matrix = np.array([query_vec, doc_vec])
    paired_matrix = sparse.csr_matrix(paired_matrix)
    
    if weight_fn and query_weight_fn:
        # nonzero() returns a tuple of two lists
        # 0 -> i dimension of non zero entries in array where i == 0 is the query vector
        # 1 -> j dimension of non zero entries in array array where j corresponds to the term_id
        document_type, term_ids = paired_matrix.nonzero()
        
        for document_type, term_id in zip(document_type, term_ids):
            if document_type == 0:
                # query stuff
                paired_matrix[document_type, term_id] = query_weight_fn(query_terms, term_id)
            else:
                # document stuff
                paired_matrix[document_type, term_id] = weight_fn(document_id, term_id)
    
    # Calculate similarity matrix
    similarity_matrix = cosine_similarity(paired_matrix)
    
    # Only first row is relevant (query document similarities)
    similarities = similarity_matrix[0, 1:]
    
    return similarities


def tfidf_query(query_terms, query_term_id):
    counts = collections.Counter(query_terms)
    
    tf = np.log(1 + counts[query_term_id])
    
    df = len(inverted_index[query_term_id])
    idf = np.log(num_documents / df)
    
    score = tf * idf
    
    return score

def tfidf(int_document_id, query_term_id):
    """
    Scoring function for a document and a query term
    
    :param int_document_id: the document id
    :param query_term_id: the query term id (assuming you have split the query to tokens)
    """
    
    # Check: RAW COUNT OR NORMALIZED TF???
    tf = np.log(1 + term_frequency(query_term_id, int_document_id))
    
    # inverted_index[query_term_id] consists of all the documents this term appears in
    df = len(inverted_index[query_term_id])
    idf = np.log(num_documents / df)
    
    score = tf * idf 
    
    return score

# Here we will normalize the raw TF and a different kind of damping: (1 + k)/k
def BM25(int_document_id, query_term_id, k=1.2, b=0.75):
    """
    Scoring function for a document and all terms of a single query
    
    :param int_document_id: the document id
    :param query_term_id: Token id of the term to be re-weighted
    
    """
    
    doc_length = len(index.document(int_document_id)[1])
    tf_raw = term_frequency(query_term_id, int_document_id)
    
    return compute_BM25(doc_length, tf_raw, query_term_id, k, b)

def BM25_query(query_terms, query_term_id, k=1.2, b=0.75):
    """
    Scoring function for a document and all terms of a single query
    
    :param int_document_id: the document id
    :param query_term_id: Token id of the term to be re-weighted
    
    """
    
    doc_length = len(query_terms) 
    counts = collections.Counter(query_terms)
    tf_raw = counts[query_term_id]
    
    return compute_BM25(doc_length, tf_raw, query_term_id, k, b)

def absolute_discounting(int_document_id, query_term_id, deltas=[0.1, 0.5, 0.9]):
    
    delta = deltas[2]
    
    # TODO: figure our if tf should be raw or not. Max to ensure tf in range [0,1]
    tf_raw = max(term_frequency(query_term_id, int_document_id) - delta, 0)
    tot_nr_terms_in_doc = len(index.document(int_document_id)[1])
    unique_terms_in_doc = unique_terms_per_document[int_document_id]
    
    corpus_size = sum(list(document_lengths.values()))
    word_frequency = collection_frequencies[query_term_id]
    p_w_c = word_frequency / corpus_size
    
    score = tf_raw / tot_nr_terms_in_doc + ((delta * tot_nr_terms_in_doc) / tot_nr_terms_in_doc) * p_w_c

    return score


def dirichlet_prior_smoothing(int_document_id, query_term_id, mu=1000):
    ''' input: one document id, one query term id, mu
        output: for one query word , the dirichlet smoothed prior
    '''

    # Calculate the components of the model: tf, |d|, p(w|c)
    tf = term_frequency(query_term_id, int_document_id)
    document_length = document_lengths[int_document_id]
    # TODO: functie van tf en p_wc maken (alhoewel de dict lookup voor TF best clean is)
    # TODO: Checken of p_wc klopt bij docenten
    p_wc = (collection_frequencies[query_term_id]/sum(collection_frequencies.values()))
#     print('tf',tf)
#     print('doc length',document_lengths[int_document_id])
#     print('p(w|C),', collection_frequencies[query_term_id],'/',sum(collection_frequencies.values()))
    dir_smoothed_prior = ((document_length/(document_length+mu))*(tf/document_length))+ ((mu/(document_length + mu))*p_wc)
    return dir_smoothed_prior

def jelinek_mercer(int_document_id, query_term_id, lambd=0.5):
    """
    Calcuate probabilty of a term by linearly interpolating with background language model.
    
    NB: Background probablity is calculated by weighting all words equally.
    
    :param int_document_id: the document id
    :param query_token_id: the query term id (assuming you have split the query to tokens)
    :param lambd: the lambda paramter for the function
    """
    
    # TODO: Checken of het klopt (was vet moe toen ik dit implementeerde)
    
    doc_length = len(index.document(int_document_id)[1])
    corpus_size = sum(collection_frequencies.values())
    
    term_freq_doc = term_frequency(query_term_id, int_doc_id)
    term_freq_corpus = sum(inverted_index[query_term_id].values())
    
    signal = (term_freq_doc / doc_length)
    background = (term_freq_corpus / corpus_size)
    
    return lambd * signal  + (1 - lambd) * background



# TODO implement tools to help you with the analysis of the results.
def PLM_score_old(int_document_id, query_id_tokens, kernel_func, mu, sigma = 50):
    '''input: one document id, one query id (so the id of one whole query)(ranging from 1-150) ,
        a kernel function, mu, sigma (=50)
    output: a language model for the document with a best position scoring strategy

    assumption: following the paper 'Positional Language Models for Information Retrieval - Yuanhua Lv, ChengXiang Zhai'
    
    Given a query, suppose all terms in a document have the
    same propagation function with the same σ, and the curve
    of the kernel density function is symmetric. Then we have
    k(i, j) = k(j, i). '''
    
    # document variables
    document = index.document(int_document_id)[1]
    doc_length = len(document)    
    
    # Query id tokens without stopwords
    query_id_no_stopwrds = [word_id for word_id in query_id_tokens if word_id > 0]
    query_length = len(query_id_no_stopwrds)
        
    # Create a matrix to store all c(w,i)' values that matter
    c_wi_prime_matrix = np.zeros((query_length,doc_length))
    
    # Todo: matrix maken van kernel
    one_sided_kernel = [kernel_func(0, j) for j in range(doc_length)]
    kernel = np.array(list(reversed(one_sided_kernel))[:-1] + one_sided_kernel)
    
    # Loop over all positions in the document
    for i, word_id_i in enumerate(document):
        # If a word occurs in the query we should propagate its frequency to other positions
        # TODO: check how to handle multiple occurences of word in query
        for q, wrd in enumerate(query_id_no_stopwrds):
            if word_id_i == wrd:
                # Loop over all positions in the document as to have a distance from position i to j
                # for positions where i is a query word
                for j in max_usefull_range(one_sided_kernel, i, doc_length):
                    c_wi_prime_matrix[q][j] += kernel[i - j]
    
    # Go through all positions in the document again now all the 
    # frequencies have been propagated
    Zi_vec = np.sum(c_wi_prime_matrix,axis=0)

    # Dirichlet smoothed probability of a word on a position in the document
    p_wc_vec = collection_freq_matrix[np.array(document)] / np.sum(collection_freq_matrix)
    smooth_pos_values = (c_wi_prime_matrix + mu * p_wc_vec) / (Zi_vec + mu)
    
    # Calculate p(w|Q)
    q_language = ml_query_word_language_model(query_id_tokens)
    
    # log( P(w|D,i) ) for all w, i pairs
    log_pos_values = np.log(smooth_pos_values)
    
    # Simplified KL-divergence formula (proportional to original for the same query)
    kl_divergence_scores = np.sum(q_language * log_pos_values.T, axis=0)
            
    maxscore = np.amax(kl_divergence_scores)
    
    return maxscore

def max_usefull_range(kernel_vector, i, N):
    
    try:
        max_distance = kernel_vector.index(0)
    except:
        max_distance = len(kernel_vector)
        
    minimum = max(0, i - max_distance)
    maximum = min(N, i + max_distance)
    
    return range(minimum, maximum)

def PLM_score(int_document_id, query_term_ids, kernel_func, mu, sigma = 50):
    '''input: one document id, one query id (so the id of one whole query)(ranging from 1-150) ,
        a kernel function, mu, sigma (=50)
    output: a language model for the document with a best position scoring strategy

    assumption: following the paper 'Positional Language Models for Information Retrieval - Yuanhua Lv, ChengXiang Zhai'
    
    Given a query, suppose all terms in a document have the
    same propagation function with the same σ, and the curve
    of the kernel density function is symmetric. Then we have
    k(i, j) = k(j, i). '''
    
    if int_document_id % 10000 == 0:
        print(int_document_id)
    
    # document variables
    document = index.document(int_document_id)[1]
    doc_length = len(document)    
    
    # Query id tokens without stopwords
    query_id_no_stopwords = [word_id for word_id in query_term_ids if word_id > 0]
    query_length = len(query_id_no_stopwords)
        
    # Create a matrix to store all c(w,i)' values that matter
#     c_wi_prime_matrix = np.zeros((query_length,doc_length))
    
    one_sided_kernel = [kernel_func(0, j) for j in range(doc_length)]
    # Add 0 pad to create even length kernel
    kernel = np.array([0] + list(reversed(one_sided_kernel))[:-1] + one_sided_kernel)
    
    i_vec = np.array(document)
    i_mat = np.tile(i_vec, query_length).reshape(query_length, doc_length)
    q_vec = np.tile(np.array(query_id_no_stopwords), doc_length).reshape(doc_length, query_length).T
    
    c_wi = i_mat == q_vec
    c_prime = np.zeros((query_length,doc_length))
    
    for i, row in enumerate(c_wi):
        # First element is edge noise
        c_prime[i] = np.convolve(kernel, row)[len(row):(1-len(row))]
    
    # Go through all positions in the document again now all the 
    # frequencies have been propagated
    Z_vec = np.sum(c_prime, axis=0)

    # Dirichlet smoothed probability of a word on a position in the document
    try:
        p_wc_vec = collection_freq_matrix[np.array(document)] / np.sum(collection_freq_matrix)
    except:
        print(np.array(document), np.array(document).dtype)
        sys.exit()
        
    smooth_pos_values = (c_prime + mu * p_wc_vec) / (Z_vec + mu)
    
    # Calculate p(w|Q)
    q_language = ml_query_word_language_model(query_term_ids)
    
    # log( P(w|D,i) ) for all w, i pairs
    log_pos_values = np.log(smooth_pos_values)
    
    # Simplified KL-divergence formula (proportional to original for the same query)
    kl_divergence_scores = np.sum(q_language * log_pos_values.T, axis=0)
            
    maxscore = np.amax(kl_divergence_scores)
    
    return int_document_id, maxscore

def quick_mats(i, N, sigma):
        normalize = np.sqrt(2*np.pi*sigma**2)
        big_area = norm.cdf( (N - i) / sigma )
        small_area = norm.cdf( (1 - i) / sigma )
        
        return normalize * (big_area - small_area)

def gaussian_kernel(i, j, sigma=50):
    return np.exp(-(i-j)**2 / (2 * sigma**2))

def triangle_kernel(i, j, sigma=50):
    return 1-abs(i - j)/sigma if abs(i - j) <= sigma else 0

def cosine_kernel(i, j, sigma=50):
    return 0.5 * (1 + np.cos((abs(i - j) * np.pi) / sigma)) if abs(i - j) <= sigma else 0

def circle_kernel(i, j, sigma=50):
    return np.sqrt(1-(abs(i - j)/sigma)**2) if abs(i - j) <= sigma else 0

def passage_kernel(i, j, sigma=50):
    return int(abs(i - j) < sigma/2)

def ml_query_word_language_model(query_term_ids):
    counts = collections.Counter(query_term_ids)
    return np.array([counts[term_id] / len(query_term_ids) for term_id in query_term_ids])

In [398]:


'''
Multinominal: p(q|lmodel_d) = prod( p(w_i | lmodel_d) )
lmodel_d: argmax p(d | lmodel_d) --> the model for which d has the highest probability of occuring
                 p(d | lmodel_d) = prod_over_w_in_doc( p(w | lmodel_d) ) (sum log)
                 log_p(d | lmodel_d) = sum_over_w_in_vocab( tf(w;d) * log_p(w | lmodel_d ) )
                 p_ml(w | lmodel_d) = tf(w;d) / len(d)
'''

def unigram_prob(term_id, document_id):
    """
    Maximum Likelihood estimate of p(word|document_model)
    """
    tf = term_frequency(term_id, document_id)
    doc_length = len(index.document(document_id)[1])
    
    return tf / doc_length

### Probabilistic score function
def log_multinomial_query_language_model(query_term_ids, document_id, word_probability_func=unigram_prob):
    """
    Query-likelihood model assuming uniform document prior
    """
    
    # sum_over_words_in_query( tf(w;d) * log( p(w|d) ) )
    return np.sum([term_frequency(document_id, term_id)*np.log(word_probability_func(document_id, term_id)) 
                   for term_id in query_term_ids])

def kl_divergenge(query_term_ids, document_id, word_probability_func=unigram_prob):
    
    p_word_given_query = query_language_model(query_term_ids)
    p_word_given_document = np.array([word_probability_func(term_id, document_id) for term_id in query_term_ids])
    
    return np.sum(p_word_given_query * np.log(p_word_given_document))

for smoothing_func in [jelinek_mercer, absolute_discounting, dirichlet_prior_smoothing]:
    print(log_multinomial_query_language_model(test_query, test_doc, smoothing_func))
    print(kl_divergenge(test_query, test_doc, smoothing_func))

-12.9976078991
-4.51789138692
-15.3273527842
-6.10855018736
-11.5034947713
-4.02351424436


In [403]:
# Figure out which queries are relevant

validation_queries_ids = []
with open('./ap_88_89/qrel_validation', 'r') as validation_queries:
    for line in validation_queries:
        query_id = line.split(' ')[0]
        if query_id not in validation_queries_ids:
            validation_queries_ids.append(query_id)
            
            

test_queries_ids = []
with open('./ap_88_89/qrel_test', 'r') as test_queries:
    for line in test_queries:
        query_id = line.split(' ')[0]
        if query_id not in test_queries_ids:
            test_queries_ids.append(query_id)

            
print('There are ',len(test_queries_ids),' test queries and ',len(validation_queries_ids),' validation queries')

There are  120  test queries and  30  validation queries


In [419]:
import pickle

def run_retrieval(model_name, score_fn, model_type, query_set):
    """
    Runs a retrieval method for all the queries and writes the TREC-friendly results in a file.
    
    :param model_name: the name of the model (a string)
    :param score_fn: the scoring function (a function - see below for an example) 
    :param model_type: indicates if a vector space or a proabilistic model is run. Takes either "Vector space"
                        or "Probabilistic"
    :param query_set: validation_queries_ids or test_queries_ids
    """
    run_out_path = '{}_even_queries.run'.format(model_name)
    pickle_path = '{}_even_queries.pickle'.format(model_name)
    if os.path.exists(run_out_path):
        print("File already exists; exiting...")
        return
    
    retrieval_start_time = time.time()

    print('Retrieving using', model_name)
    

    # The dictionary data should have the form: query_id --> (document_score, external_doc_id)
    data = {}
    if model_name == 'TF-IDF':
        weight_fn = tfidf
        query_weight_fn = tfidf_query
    if model_name == 'BM25':
        weight_fn = BM25
        query_weight_fn = BM25_query
    counter = 1

    for query_id, query_words in tokenized_queries.items():
        # Run for uneven queries in query set
        if int(query_id) % 2 == 1:
            continue
        if query_id not in query_set:
            continue

        for int_doc_id in range(index.document_base(), index.maximum_document()):
            ext_doc_id, doc_token_ids = index.document(int_doc_id)
            # Only calculate score of document if it contains query words,
            # otherwise the score is zero for the vector based models
            length_query = len(query_words)
            for i, query_word in enumerate(query_words):
                # Only compute score if the query word is in the document
                if int_doc_id in inverted_index[query_word].keys():

                    
                    # Calculate score for document and query terms
                    if model_type == "Vector space":
                        score = score_similarity(query_words, int_doc_id, weight_fn, query_weight_fn)
                    elif model_type == "Probabilistic":
                        score = score_fn(query_words, int_doc_id)

                    if query_id in data.keys():
                        data[query_id].append(tuple([float(score), str(ext_doc_id)]))
                    else:
                        data[query_id] = [((tuple([float(score),str(ext_doc_id)])))]

                    # If the document has a query word, than the score is calculated and appended by this point
                    break


                # If up until the last query word, the query word did not appear in the 
                # document, the score is 0.0 
                elif i == length_query-1:
                    score = 0.0
                    if query_id in data.keys():
                        data[query_id].append(tuple([float(score), str(ext_doc_id)]))
                    else:
                        data[query_id] = [((tuple([float(score),str(ext_doc_id)])))]



        print('queries: ', counter,'/',len(query_set)/2,'\t this took: ', time.time() - retrieval_start_time, ' seconds in total' )
        counter += 1

        # transform the list of tuples to a tuple of tuples, since this is the requered datastructue
        data[query_id] = tuple(data[query_id])   

    # Overwrite the previous pickle and .run file with the new (accumulated) data variable
    with open(pickle_path, 'wb') as f_pickle:
        pickle.dump(data, f_pickle, protocol=pickle.HIGHEST_PROTOCOL)

    with open(run_out_path, 'w') as f_out:
        write_run(
            model_name=model_name,
            data=data,
            out_f=f_out,
            max_objects_per_query=1000)

In [421]:
# run run_retrieval

'''
Jelinek-Mercer (explore different values of 𝛌 in the range [0.1, 0.5, 0.9]). [5 points]
Dirichlet Prior (explore different values of 𝛍 [500, 1000, 1500]). [5 points]
Absolute discounting (explore different values of 𝛅 in the range [0.1, 0.5, 0.9]). [5 points]
'''

data={
        'Q1': ((1.0, 'DOC1'), (0.5, 'DOC2'), (0.75, 'DOC3')),
        'Q2': ((-0.1, 'DOC1'), (1.25, 'DOC2'), (0.0, 'DOC3')),
    }

# jelinek_01 = (functools.partial(jelinek_mercer, lambd=0.1), 'jelinek_0.1')
jelinek_05 = (functools.partial(jelinek_mercer, lambd=0.5), 'jelinek_0.5')
jelinek_09 = (functools.partial(jelinek_mercer, lambd=0.9), 'jelinek_0.9')
dirichlet_500 = (functools.partial(dirichlet_prior_smoothing, mu=500), 'dirichlet_500')
dirichlet_1000 = (functools.partial(dirichlet_prior_smoothing, mu=1000), 'dirichlet_1000')
dirichlet_1500 = (functools.partial(dirichlet_prior_smoothing, mu=1500), 'dirichlet_1500')
absolute_01 = (functools.partial(absolute_discounting, delta=0.1), 'absolute_0.1')
absolute_05 = (functools.partial(absolute_discounting, delta=0.5), 'absolute_0.5')
absolute_09 = (functools.partial(absolute_discounting, delta=0.9), 'absolute_0.9')

smoothings_functions = [jelinek_05, jelinek_09, dirichlet_500, dirichlet_1000, dirichlet_1500
                       absolute_01, absolute_05, absolute_09]

import cProfile, pstats
from io import StringIO
pr = cProfile.Profile()
pr.enable()

for smoothing_func, label in smoothings_functions:
    run_retrieval(label, 
                  functools.partial(log_multinomial_query_language_model, word_probability_func=smoothing_func), 
                  "Probabilistic", validation_queries_ids)

pr.disable()
s = StringIO()
sortby = 'cumulative'
ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
ps.print_stats()
print(s.getvalue())

Retrieving using jelinek_0.1
queries:  1 / 15.0 	 this took:  14.140056848526001  seconds in total
queries:  2 / 15.0 	 this took:  20.679800748825073  seconds in total
queries:  3 / 15.0 	 this took:  32.89985489845276  seconds in total
queries:  4 / 15.0 	 this took:  66.32235169410706  seconds in total
queries:  5 / 15.0 	 this took:  150.22247672080994  seconds in total
queries:  6 / 15.0 	 this took:  162.29551577568054  seconds in total
queries:  7 / 15.0 	 this took:  184.56936502456665  seconds in total
queries:  8 / 15.0 	 this took:  244.23482394218445  seconds in total
queries:  9 / 15.0 	 this took:  391.52746176719666  seconds in total
queries:  10 / 15.0 	 this took:  508.76024985313416  seconds in total
queries:  11 / 15.0 	 this took:  518.0915699005127  seconds in total
queries:  12 / 15.0 	 this took:  538.4126689434052  seconds in total
queries:  13 / 15.0 	 this took:  567.9230217933655  seconds in total
queries:  14 / 15.0 	 this took:  580.007533788681  seconds in

In [83]:
tokenized_queries.items()

dict_items([('51', [5872, 3066]), ('52', [54, 846, 2208]), ('53', [4886, 48553]), ('54', [2231, 1339, 1490]), ('55', [4400, 235]), ('56', [389, 4081, 241, 2294, 190739]), ('57', [155285]), ('58', [3454, 1920]), ('59', [804, 955, 90480]), ('60', [157422, 213, 215481]), ('61', [358, 523, 335, 1098, 2598]), ('62', [55, 64853, 73364]), ('63', [2129, 243152]), ('64', [1987, 451]), ('65', [281, 202871, 982]), ('66', [1535, 1434, 3628]), ('67', [3698, 5802, 416, 76611]), ('68', [248, 111208, 1322, 74181, 91961]), ('69', [2354, 203193, 3177, 811, 942]), ('70', [233033, 163926]), ('71', [580, 120650]), ('72', [71970, 218232, 5465, 276]), ('73', [71970, 218232, 21, 43984]), ('74', [61863, 238]), ('75', [30473]), ('76', [5465, 276, 1633, 1704, 3895]), ('77', [188098]), ('78', [105220]), ('79', [96348, 78, 33, 1801]), ('80', [214, 252, 789, 187515]), ('81', [380, 66582, 237310, 3367, 3383, 1893]), ('82', [5967, 3067]), ('83', [1122, 1189, 2783]), ('84', [2563, 201208, 601, 351, 851, 122153]), ('85

In [33]:
# Document and query share terms
test_query = tokenized_queries['52']
test_doc = 11

print(test_query)
print(tokenized_queries.items())
print(score_similarity(test_query, test_doc, weight_fn=tfidf, query_weight_fn=tfidf_query))
print(score_similarity(test_query, test_doc, weight_fn=BM25, query_weight_fn=BM25_query))
print(score_similarity(test_query, test_doc))

[54, 846, 2208]
dict_items([('51', [5872, 3066]), ('52', [54, 846, 2208]), ('53', [4886, 48553]), ('54', [2231, 1339, 1490]), ('55', [4400, 235]), ('56', [389, 4081, 241, 2294, 190739]), ('57', [155285]), ('58', [3454, 1920]), ('59', [804, 955, 90480]), ('60', [157422, 213, 215481]), ('61', [358, 523, 335, 1098, 2598]), ('62', [55, 64853, 73364]), ('63', [2129, 243152]), ('64', [1987, 451]), ('65', [281, 202871, 982]), ('66', [1535, 1434, 3628]), ('67', [3698, 5802, 416, 76611]), ('68', [248, 111208, 1322, 74181, 91961]), ('69', [2354, 203193, 3177, 811, 942]), ('70', [233033, 163926]), ('71', [580, 120650]), ('72', [71970, 218232, 5465, 276]), ('73', [71970, 218232, 21, 43984]), ('74', [61863, 238]), ('75', [30473]), ('76', [5465, 276, 1633, 1704, 3895]), ('77', [188098]), ('78', [105220]), ('79', [96348, 78, 33, 1801]), ('80', [214, 252, 789, 187515]), ('81', [380, 66582, 237310, 3367, 3383, 1893]), ('82', [5967, 3067]), ('83', [1122, 1189, 2783]), ('84', [2563, 201208, 601, 351, 851

In [396]:
import cProfile, pstats
from io import StringIO
import functools

print([id2token[x] for x in test_query if x > 0 ])

PLM_triangle = functools.partial(PLM_score_old, kernel_func=triangle_kernel, mu=1000)

pr = cProfile.Profile()
pr.enable()

print(PLM_triangle(test_doc, test_query))

pr.disable()
s = StringIO()
sortby = 'cumulative'
ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
ps.print_stats()
print(s.getvalue())

['south', 'african', 'sanctions']
-inf
         1592 function calls in 0.005 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    0.005    0.002 /Volumes/Macintosh_HD/Users/Aron/miniconda3/envs/ml1labs/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2832(run_code)
        2    0.000    0.000    0.004    0.002 {built-in method builtins.exec}
        1    0.000    0.000    0.004    0.004 <ipython-input-396-884e48b46823>:12(<module>)
        1    0.001    0.001    0.004    0.004 <ipython-input-371-2cf675f70044>:174(PLM_score_old)
        3    0.000    0.000    0.001    0.000 /Volumes/Macintosh_HD/Users/Aron/miniconda3/envs/ml1labs/lib/python3.6/site-packages/numpy/core/fromnumeric.py:1730(sum)
        4    0.001    0.000    0.001    0.000 {method 'reduce' of 'numpy.ufunc' objects}
        3    0.000    0.000    0.001    0.000 /Volumes/Macintosh_HD/Users/Aron/miniconda3/envs/ml1labs/lib/

/Volumes/Macintosh_HD/Users/Aron/miniconda3/envs/ml1labs/lib/python3.6/site-packages/ipykernel_launcher.py:224: RuntimeWarning: divide by zero encountered in log


In [ ]:
# Document and query share terms
test_query = tokenized_queries['52']
test_doc = 11

%time print(score_similarity(test_query, test_doc, weight_fn=tfidf, query_weight_fn=tfidf_query))
%time print(score_similarity(test_query, test_doc, weight_fn=BM25, query_weight_fn=BM25_query))
%time print(score_similarity(test_query, test_doc))

In [ ]:
ext_doc_id, doc_token_ids = index.document(1)

# This contains the length of each document in the collection.
document_lengths

# This contains the number of unique terms per document in the collection.
unique_terms_per_document

# This contains the frequency in which the query term appears in all the documents
collection_frequencies[35]

# This contains the frequency in which the query term appears in a given document
inverted_index[880]

collection_frequencies[35]

### Task 2: Latent Semantic Models (LSMs) [15 points] ###

In this task you will experiment with applying distributional semantics methods ([LSI](http://lsa3.colorado.edu/papers/JASIS.lsi.90.pdf) **[5 points]** and [LDA](https://www.cs.princeton.edu/~blei/papers/BleiNgJordan2003.pdf) **[5 points]**) for retrieval.

You do not need to implement LSI or LDA on your own. Instead, you can use [gensim](http://radimrehurek.com/gensim/index.html). An example on how to integrate Pyndri with Gensim for word2vec can be found [here](https://github.com/cvangysel/pyndri/blob/master/examples/word2vec.py). For the remaining latent vector space models, you will need to implement connector classes (such as `IndriSentences`) by yourself.

In order to use a latent semantic model for retrieval, you need to:
   * build a representation of the query **q**,
   * build a representation of the document **d**,
   * calculate the similarity between **q** and **d** (e.g., cosine similarity, KL-divergence).
     
The exact implementation here depends on the latent semantic model you are using. 
   
Each of these LSMs come with various hyperparameters to tune. Make a choice on the parameters, and explicitly mention the reasons that led you to these decisions. You can use the validation set to optimize hyper parameters you see fit; motivate your decisions. In addition, mention clearly how the query/document representations were constructed for each LSM and explain your choices.

In this experiment, you will first obtain an initial top-1000 ranking for each query using TF-IDF in **Task 1**, and then re-rank the documents using the LSMs. Use TREC Eval to obtain the results and report on `NDCG@10`, Mean Average Precision (`MAP@1000`), `Precision@5` and `Recall@1000`.

Perform significance testing **[5 points]** (similar as in Task 1) in the class of semantic matching methods.

In [ ]:
import gensim.models

    
# Under construction:
def LSI(corpus, dictionary, num_topics):
    return gensim.models.lsimodel.LsiModel(corpus, id2word=dictionary, num_topics=num_topics)

# Pending:
def LDA(arg1, arg2):
    return gensim.models.ldamodel.LdaModel(corpus=sentences, id2word=words, num_topics=10)

In [ ]:
# This is the connector that ports Pyndri objects in a Gensim-friendly way.
# Source: http://www.pythonexample.com/code/gensim-pyro/
# Only updated return values from _doc2bow func; swapped key, value

class LDALSISentences(gensim.interfaces.CorpusABC):
 
    def __init__(self, index, dictionary, max_documents=None):
        assert isinstance(index, pyndri.Index)
 
        self.index = index
        self.dictionary = dictionary
 
        self.max_documents = max_documents
 
    def _maximum_document(self):
        if self.max_documents is None:
            return self.index.maximum_document()
        else:
            return min(
                self.max_documents + self.index.document_base(),
                self.index.maximum_document())
 
    def _doc2bow(self, doc):
 
        di = collections.defaultdict(int)
 
        for token_id in doc:
            di[token_id] += 1
    
        return [(key, value) for key, value in di.items()]
 
    def __iter__(self):

        for int_doc_id in range(self.index.document_base(),
                                self._maximum_document()):
            ext_doc_id, tokens = self.index.document(int_doc_id)
 
            tokens = tuple(
                token_id
                for token_id in tokens
                if token_id > 0 and token_id in self.dictionary)
 
            yield self._doc2bow(tokens)
 
    def __len__(self):
        return self._maximum_document() - self.index.document_base()

In [ ]:
# Import dependencies:
import gensim
import pickle
from scipy.sparse import dok_matrix
from scipy import spatial

# Step 0: Convert corpus into gensim dictionary object
def init_corpus(store_loc_name='/tmp/tf_corpus.mm'):
    
    #index = pyndri.Index('index/')
    print("Building dictionary...")
    dictionary = pyndri.extract_dictionary(index)
    
    print("Building corpus...")
    corpus = LDALSISentences(index, dictionary)
    
    print("Building TF-Corpus...")
    corpus = [doc for doc in corpus]
    
    print("Storing dictionary and corpus...")
    #dictionary.save('index.dict')
    gensim.corpora.MmCorpus.serialize(store_loc_name, corpus)

    return corpus, dictionary, store_loc_name
    
def load_corpus(store_loc_name):
    return gensim.corpora.MmCorpus(store_loc_name)

def transform_corpus(corpus):
    tfidf_model = gensim.models.TfidfModel(corpus)
    corpus_tfidf = tfidf_model[corpus]
    return corpus_tfidf

# Step 1: Build a representation q of the query
def query_representation(dictionary):
    
    with open('./ap_88_89/topics_title', 'r') as f_topics:
        queries = parse_topics([f_topics])
        
    tokenized_queries = {
        query_id: [dictionary.translate_token(token)
                   for token in index.tokenize(query_string)
                   if dictionary.has_token(token)]
        for query_id, query_string in queries.items()}

    unique_query_ids = set(
        query_term_id
        for query_term_ids in tokenized_queries.values()
        for query_term_id in query_term_ids)
    
    query_bows = [dictionary.doc2bow(query)
                  for query in tokenized_queries.values()]
    
    return tokenized_queries, unique_query_ids, query_bows

# Step 2: Build a representation d of the documents i.e. by scoring them
def document_representation(index, dictionary, method="tfidf"):
    """ Returns a documents representation; """
    X = dok_matrix((len(dictionary), len(index) + 1), dtype=np.float32)
    start_time = time.time()
    # TODO: You only care about words occurring in the query, since otherwise IF-IDF == 0 (?)

    if method == "tfidf":
        for query_id in inverted_index.keys():
            for doc_id in inverted_index[query_id]:
                X[query_id, doc_id] = tfidf(doc_id, query_id)

    elif method == "tf":
        for query_id in inverted_index.keys():
            for doc_id in inverted_index[query_id]:
                X[query_id, doc_id] = term_frequency(query_id, doc_id)
    
    return X

# Step 3: Build a function that is able to calculate similarity between q and d.
def cosine_similarity(vector_a, vector_b):
    
    #vector_a = X[:,1].toarray().flatten()
    #vector_b = X[:,2].toarray().flatten()
    
    return 1 - spatial.distance.cosine(vector_a, vector_b)

# Generates a weird error:
def gensim_cosine_similarity(model, corpus, query_bow, save=False, store_name=""):
    
    sim_index = gensim.similarities.MatrixSimilarity(model[corpus]) 
    
    query_lsi = lsi[query_bow]
    
    if save == True:
        sim_index.save(store_name)
    
    similarities = sim_index[query_lsi]
    
    return similarities

# Step 4: Use the representations to train LSMs - choose hyperparams by testing


# Step 5: Use TF-IDF to rank top 1000 documents, for each query

# Step 6: Use both LSMs to re-rank top 1000 documents and apply TREC to test results. Based on:
## - NDCG@10
## - MAP@1000
## - Precision@5
## - Recall@1000

%time corpus, dictionary, storage_loc = init_corpus()

In [ ]:
%time _, _, query_bows = query_representation(dictionary)
pickle.dump(query_bows, open( "query_bows.p", "wb" ))

In [ ]:
with open('query_bows.p', 'rb') as file:
    test = pickle.load(file)

In [ ]:
lsi_model = LSI(corpus,dictionary, 300)
pickle.dump(lsi_model, open("lsi_model.p", "wb"))

In [ ]:
%time corpus_lsi = lsi[corpus]

sim_index = gensim.similarities.MatrixSimilarity(lsi_model[corpus_lsi])

#query_bow = query_bows[0]

#query_lsi = lsi[query_bow]

#similarities = sim_index[query_lsi]


In [ ]:
# Own implementation of connector
import pyndri.compat

# Note: rerunning line 5 may throw internal errors... Not sure why.
index = pyndri.Index('index/')

# Builds the dictionary:
dictionary = pyndri.extract_dictionary(index)

# Builds an iterable object for computational- and memory efficiency:
sentences = pyndri.compat.IndriSentences(index, dictionary)

print("Number of unique terms / Dictionary: ", len(sentences.dictionary))
print("Number of documents: ", sentences.index)

corpus = []

for doc in range(index.document_base(), index.maximum_document()):
    corpus.append(dictionary.doc2bow(index.document(doc)[1]))

for i, doc in enumerate(corpus):
    print(doc)
    
    # Print only first 5 document representations:
    if i == 5:
        break
        
gensim.corpora.MmCorpus.serialize('/tmp/simple_bow.mm', corpus)

corpus = gensim.corpora.MmCorpus('/tmp/simple_bow.mm')

# This contains a bug; "KeyError: 0", will be fixed next commit...
%time lsi = gensim.models.LsiModel(corpus, id2word=dictionary, num_topics=300)

In [ ]:
# Gensim tutorial to see how it works, no Pyndri included:

texts = [['human', 'interface', 'computer'],
         ['survey', 'user', 'computer', 'system', 'response', 'time'],
         ['eps', 'user', 'interface', 'system'],
         ['system', 'human', 'system', 'eps'],
         ['user', 'response', 'time'],
         ['trees'],
         ['graph', 'trees'],
         ['graph', 'minors', 'trees'],
         ['graph', 'minors', 'survey']]

dictionary = gensim.corpora.Dictionary(texts)
print(dictionary)

token2id = dictionary.token2id
print(token2id)

corpus = [dictionary.doc2bow(text) for text in texts]
print("Here", corpus)

gensim.corpora.MmCorpus.serialize('/tmp/sample.mm', corpus)

corpus = gensim.corpora.MmCorpus('/tmp/sample.mm')
print(corpus)

tfidf = gensim.models.TfidfModel(corpus)
print(tfidf)

corpus_tfidf = tfidf[corpus]
print(corpus_tfidf)

lsi = gensim.models.LsiModel(corpus, id2word=dictionary, num_topics=5)
print(lsi.print_topics(5))


### Task 3:  Word embeddings for ranking [20 points] (open-ended) ###

First create word embeddings on the corpus we provided using [word2vec](http://arxiv.org/abs/1411.2738) -- [gensim implementation](https://radimrehurek.com/gensim/models/word2vec.html). You should extract the indexed documents using pyndri and provide them to gensim for training a model (see example [here](https://github.com/nickvosk/pyndri/blob/master/examples/word2vec.py)).
   
This is an open-ended task. It is left up you to decide how you will combine word embeddings to derive query and document representations. Note that since we provide the implementation for training word2vec, you will be graded based on your creativity on combining word embeddings for building query and document representations.

Note: If you want to experiment with pre-trained word embeddings on a different corpus, you can use the word embeddings we provide alongside the assignment (./data/reduced_vectors_google.txt.tar.gz). These are the [google word2vec word embeddings](https://code.google.com/archive/p/word2vec/), reduced to only the words that appear in the document collection we use in this assignment.

### Task 4: Learning to rank (LTR) [15 points] (open-ended) ###

In this task you will get an introduction into learning to rank for information retrieval.

You can explore different ways for devising features for the model. Obviously, you can use the retrieval methods you implemented in Task 1, Task 2 and Task 3 as features. Think about other features you can use (e.g. query/document length). Creativity on devising new features and providing motivation for them will be taken into account when grading.

For every query, first create a document candidate set using the top-1000 documents using TF-IDF, and subsequently compute features given a query and a document. Note that the feature values of different retrieval methods are likely to be distributed differently.

You are adviced to start some pointwise learning to rank algorithm e.g. logistic regression, implemented in [scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).
Train your LTR model using 10-fold cross validation on the test set. More advanced learning to rank algorithms will be appreciated when grading.

### Task 4: Write a report [15 points; instant FAIL if not provided] ###

The report should be a PDF file created using the [sigconf ACM template](https://www.acm.org/publications/proceedings-template) and will determine a significant part of your grade.

   * It should explain what you have implemented, motivate your experiments and detail what you expect to learn from them. **[10 points]**
   * Lastly, provide a convincing analysis of your results and conclude the report accordingly. **[10 points]**
      * Do all methods perform similarly on all queries? Why?
      * Is there a single retrieval model that outperforms all other retrieval models (i.e., silver bullet)?
      * ...

**Hand in the report and your self-contained implementation source files.** Only send us the files that matter, organized in a well-documented zip/tgz file with clear instructions on how to reproduce your results. That is, we want to be able to regenerate all your results with minimal effort. You can assume that the index and ground-truth information is present in the same file structure as the one we have provided.
